rust/azure_iot_operations_services/src/leased_lock/lease.rs

```diff
@@ -0,0 +1,416 @@
+    ///
+    /// [`struct@Error`] of kind [`ServiceError`](ErrorKind::ServiceError) if the State Store returns an Error response
+    ///
+    /// [`struct@Error`] of kind [`UnexpectedPayload`](ErrorKind::UnexpectedPayload) if the State Store returns a response that isn't valid for a `Set` request
+    ///
+    /// [`struct@Error`] of kind [`AIOProtocolError`](ErrorKind::AIOProtocolError) if there are any underlying errors from the command invoker
+    ///
+    /// [`struct@Error`] of kind [`KeyAlreadyLeased`](ErrorKind::KeyAlreadyLeased) if the `lease` is already in use by another holder
+    /// # Panics
+    /// Possible panic if, for some error, the fencing token (a.k.a. `version`) of the acquired lease is None.
+    pub async fn try_acquire(
+        &mut self,
+        lease_expiration: Duration,
+        request_timeout: Duration,
+    ) -> Result<HybridLogicalClock, Error> {
+        let acquire_result = Self::internal_try_acquire(
+            self.state_store.clone(),
+            self.key_name.clone(),
+            self.lease_holder_name.clone(),
+            lease_expiration,
+            request_timeout,
+            self.current_fencing_token.clone(),
+        )
+        .await;
+
+        match acquire_result {
+            Ok(ref version) => {
+                *self.current_fencing_token.clone().lock().await = Some(version.clone());
+            }
+            Err(_) => {
+                *self.current_fencing_token.clone().lock().await = None;
+            }
+        }
+
+        acquire_result
+    }
+
+    fn start_lease_auto_renewal(
+        state_store: Arc<state_store::Client<C>>,
+        key_name: Vec<u8>,
+        lease_holder_name: Vec<u8>,
+        lease_expiration: Duration,
+        request_timeout: Duration,
+        renewal_period: Option<Duration>,
+        current_fencing_token: Arc<Mutex<Option<HybridLogicalClock>>>,
+    ) -> JoinHandle<()> {
+        match renewal_period {
+            Some(period) => {
+                tokio::task::spawn({
+                    async move {
+                        let mut renewal_interval = tokio::time::interval(period);
+
+                        loop {
+                            renewal_interval.tick().await;
+
+                            match Self::internal_try_acquire(
+                                state_store.clone(),
+                                key_name.clone(),
+                                lease_holder_name.clone(),
+                                lease_expiration,
+                                request_timeout,
+                                current_fencing_token.clone(),
+                            )
+                            .await
+                            {
+                                Ok(_) => {
+                                    // new fencing token is already saved by `internal_try_acquire()`.
+                                }
+                                Err(_) => {
```

Comment:
Since you only want to do something on `Err` perhaps you should use `is_err()` instead of `match`

Reply:
Code has been reworked, could you please review again? Thanks.